STP Refactor 2022

kfu02 opened the pull request:
## Description
Implements gameplay changes, including 1) greedy role assignment, 2) redefinition of Tactics and Roles, and 3) simplification of existing architecture. See design doc below for more details.

## Design Documents
[STP Refactor 2022](https://docs.google.com/document/d/1gRAF--W7FwGoof0--1l_pyjM-N4RmMiV6FhBvyQIRgM/edit?usp=sharing)

## Steps to test
Launch sim. Run different plays by changing the `test_play` variable at the bottom of `gameplay_node.py`. (The only three plays written so far are the ones imported [at the top of that file](https://github.com/RoboJackets/robocup-software/pull/1811/files#diff-90a35bbaea2fb27fd2d7efbdd3896b7a09d65ca6b7054300c7df727e3b00ac02R33).) Ensure that the plays behave as their docstrings would suggest.

Note: sometimes world_state is not passed to gameplay correctly and everything breaks. Try launching a few times. The reason for this bug is unclear; I added a task to the board to address this.

## Other Review Notes

I intentionally left the unused imports at the top of gameplay_node and the "hanging indent" complaints from pylint. Unused imports are there so you know which plays are testable. Hanging indent was left to be fixed by changing either black auto-formatter or pylint preferences.

There are a lot of files in this PR. Here's one _suggested_ way to read them:

 1. SRTP structure: `stp/skill/__init__, stp/role/__init__, stp/tactic/__init__, stp/play/__init__, stp/role/cost.py, gameplay_node.py`
 2. line_up play: `play/line_up, tactic/line_tactic, role/dumb_move, skill/move`
 3. basic_defense play: `play/basic_defense, tactic/goalie_tactic, tactic/wall_tactic, role/goalie_rolewall_calculations` + pivot_kick skill (below)
 4. pivot_kick skill: `skill/capture, skill/kick, skill/pivot, skill/pivot_kick`
 5. keepaway play: `play/keepaway, tactic/pass_tactic, role/passer, role/receiver, skill/receive, skill/settle`
 6. least important files: `coordinator, move_tactic, role/__init__` (all currently unused or empty), `rc.py` (literally deleted one whitespace line), `stp/role/assignment/__init__, stp/situation/__init__` (typing changes), `basic_play_selector` (only ~10 of the 97 lines are actual logic changes, mostly formatting, and the logic change is dead simple)

Comment or Slack with any concerns!

kfu02 committed:
change plays to be controllers of role assignment, redefine STP interfaces, rm prev_result everywhere

kfu02 committed:
Merge branch 'ros2' into role-assignment-update-2022

kfu02 committed:
fix style errors from last commit

kfu02 committed:
add basic lineUp play to showcase new stp, brick old gameplay system

kfu02 committed:
make pretty-lines

kfu02 committed:
clean up lineup play

kfu02 committed:
Merge branch 'stp-cleanup' into role-assignment-update-2022

kfu02 committed:
move costfns to common file, delete old costfns, show different costfn use in lineup play

kfu02 committed:
merge ordered costs and roles into one list of (new) RoleRequests

kfu02 committed:
minor cleanup and make pretty lines

kfu02 committed:
apply new stp to basic defense (still rocky), fix corresponding skills/tactics

kfu02 committed:
adjust tactics to allow multi-robot tactic

kfu02 committed:
continue multi-robot tactics

kfu02 committed:
Revert "continue multi-robot tactics"

This reverts commit a58007affdedecc9158ec6f18fca4fdedebeb9a2.

kfu02 committed:
Revert "adjust tactics to allow multi-robot tactic"

This reverts commit 669873f384d4afd8d82a5304e1bd9aaa27172723.

kfu02 committed:
change to asrtp framework by making Role class and Tactics hold Roles

kfu02 committed:
working wall tactic

kfu02 committed:
add goalie role/tactic to basic def, make pretty-lines

kfu02 committed:
fix style errors, add comments/typing

kfu02 committed:
Merge branch 'ros2' into role-assignment-update-2022

kfu02 committed:
write half of pass tactic

kfu02 committed:
fix cap isdone

kfu02 committed:
working pass except planner node crash?"

kfu02 committed:
fix bug by lowering kick speed?"

kfu02 commented:
correction: planner bug happens when ball is placed on left of field for passer, and receiver is moving on right of field. no idea why. something to do with coordinates?

regardless passing works when ball starts not on left, suggesting it's a planning coord bug

EDIT: see #1828

kfu02 commented on rj_gameplay/stp/coordinator.py:
```diff
@@ -27,16 +27,16 @@ class Coordinator:
-    _props: Dict[Type[stp.play.IPlay], Any]
+    _props: Dict[Type[stp.play.Play], Any]
 
     # TODO(1585): Properly handle type annotations for props instead of using Any.
```
delete this todo (no props)

Yudai-8-om committed:
apply hash

kfu02 committed:
rm stray todo (mostly to re-run checks)

kfu02 committed:
clean up role assignment, discover bug and leave todo

kfu02 committed:
make keepaway play pass indefinitely

kfu02 committed:
make pretty-lines

Yudai-8-om committed:
merge and fix

Yudai-8-om committed:
delete TODO

N8BWert committed:
added an update call to self.robot in the tick method of the Skill superclass

N8BWert committed:
changed self.robot update references in capture.py

N8BWert committed:
Changed a type I added by accident

gordonm867 committed:
extremely sketchy yet somewhat magical fix

gordonm867 committed:
Undid last commit

I meant to make a separate branch and open a PR, but instead I accidentally committed to this branch.  There was only one file changed and I pasted the original back in.  This branch should now be no different than how it was before I started and I will make a PR.

gordonm867 committed:
magic fixes to trajectory splicing issue

kfu02 committed:
rm coordinator from gameplay_node, as plays return intents directly now

kfu02 committed:
rm TODO ab world_state none as it already does that

kfu02 committed:
do what last commit said; did not make correct edit

kfu02 committed:
clean up dumb move role

kfu02 committed:
add docstrings for abstract tactic class

kfu02 committed:
add docstrings for play; make pretty

kfu02 committed:
add docs for role

kfu02 committed:
add todos to line tactic

kfu02 committed:
update receive/cap/setl docs

kfu02 committed:
update dumb move doc

kfu02 committed:
update goalie tactic doc

kfu02 committed:
update play docs

kfu02 committed:
update more docs

kfu02 commented on rj_gameplay/rj_gameplay/basic_play_selector.py:
```diff
@@ -19,46 +19,47 @@
 
 # TODO: Put new plays into the dict properly
```
not sure what this TODO meant so I left it there

HussainGynai replied:
Basically saying that we should have a way to change what plays are selectable on the fly. Like maybe we have a more aggressive playbook if we're down at the half, or maybe a really defensive one if we have the lead.

kfu02 commented on rj_gameplay/rj_gameplay/calculations/wall_calculations.py:
```diff
@@ -41,7 +40,7 @@ def find_wall_pts(num_wallers: int, world_state: rc.WorldState) -> List[np.ndarr
 
     # get direction vec
-    dir_vec = (ball_pt - goal_pt) / np.linalg.norm(ball_pt - goal_pt)
+    dir_vec = (ball_pt - goal_pt) / (np.linalg.norm(ball_pt - goal_pt) + 1e-9)
```
this file will be removed by #1821

kfu02 commented on rj_gameplay/rj_gameplay/gameplay_node.py:
```diff
@@ -60,10 +45,15 @@ class GameplayNode(Node):
         self,
         play_selector: situation.IPlaySelector,
+        test_play: Optional[stp.play.Play] = None,
+        # TODO: see whether or not world_state is ever not None here
```
I have a hunch that gameplay_node is never given a non-null world_state on init. Haven't tested yet, so I left this as a TODO.

HussainGynai replied:
It just depends on boot order of the ROS node (which is pretty random), I don't think its really an issue.

kfu02 commented on rj_gameplay/rj_gameplay/gameplay_node.py:
```diff
@@ -463,11 +459,10 @@ def shutdown(self) -> None:
     play_selector = basic_play_selector.BasicPlaySelector()
 
-    gameplay = GameplayNode(play_selector)
+    # change this line to test different plays
```
```suggestion
    # change this line to test different plays (set to None if no desired test play)
```

kfu02 commented on rj_gameplay/rj_gameplay/play/basic_defense.py:
```diff
@@ -1,83 +1,43 @@
+    """Play that consists of:
+     - 1 Goalie
+     - 5 Wallers
+     TODO: add 2 aggressive markers, go down to 3 Wallers
```
@Yudai-8-om is handling this: https://app.clickup.com/t/27b6uh5

kfu02 commented on rj_gameplay/rj_gameplay/play/basic_defense.py:
```diff
@@ -1,83 +1,43 @@
-    def compute_props(self, prev_props):
-        pass
+        # super simple FSM
+        # TODO: use FSM class (or at least don't use string literals)
```
I used this hacky string-based FSM style to test a lot of the new architecture quickly. The correct way to do this is to have an FSM class, and give it the states and transitions between them. There's an FSM class in util/fsm.py that may suit our needs, or it may be overcomplicated; I haven't yet looked.

Another option, as stated in the design doc, is to go with FSM for important interrupts (e.g. PassTactic) and use a behavior tree/utility theory/some other control system to handle generic behavior.

kfu02 commented on rj_gameplay/rj_gameplay/role/goalie_role.py:
```diff
@@ -0,0 +1,161 @@
+
+
+def get_goalie_pt(world_state: stp.rc.WorldState) -> np.ndarray:
+    """Finds point for goalie to best be in to block a shot.
```
```suggestion
    """Gives goalie a default location to track the ball from when it is not actively intercepting or capturing the ball.
```

kfu02 commented on rj_gameplay/rj_gameplay/role/goalie_role.py:
```diff
@@ -0,0 +1,161 @@
+    return mid_pt
+
+
+def get_block_pt(world_state: stp.rc.WorldState, my_pos: np.ndarray) -> np.ndarray:
```
```suggestion
"""Calculates point to move to to stop the ball, standing in for an intercept planner.
"""
def get_block_pt(world_state: stp.rc.WorldState, my_pos: np.ndarray) -> np.ndarray:
```

kfu02 commented on rj_gameplay/rj_gameplay/role/goalie_role.py:
```diff
@@ -0,0 +1,161 @@
+        self.pivot_kick_skill = None
+
+    # TODO: there is crusty if-else here, use utility AI or behavior tree or FSM
+    #       anything more readable than this (behavior tree prob best fit for current logic)
```
addressed above

kfu02 commented on rj_gameplay/rj_gameplay/role/passer.py:
```diff
@@ -0,0 +1,67 @@
+         - when got ball: mark pass ready for Tactic, dribble, wait
+         - on pass signal from Tactic: pivot_kick to point, let receiver get ball, done
+        """
+        # print("passer state:", self._state)
```
```suggestion
```

kfu02 commented on rj_gameplay/rj_gameplay/role/receiver.py:
```diff
@@ -0,0 +1,49 @@
+         - interrupt signal from Tactic: go get ball
+         - when got ball: done
+        """
+        # print("receiver state:", self._state)
```
```suggestion
```

kfu02 commented on rj_gameplay/rj_gameplay/skill/capture.py:
```diff
@@ -5,44 +5,66 @@
+        # this doesn't work:
+        # robot_pos = self.robot.pos
+        # because self.robot is passed on init and never updated
+        # TODO: make superclass force robot update in tick()
```
addressed by #1824

kfu02 commented on rj_gameplay/rj_gameplay/skill/pivot_kick.py:
```diff
@@ -36,6 +36,7 @@ def __init__(
 
         self.__name__ = "pivot kick"
         self.robot = robot
+        # TODO: make skill take either np array or tuple (and auto-cast to whichever is best)
```
addressed by @Rich-Hoepfinger : https://app.clickup.com/t/1ydy7cd

kfu02 commented on rj_gameplay/rj_gameplay/tactic/line_tactic.py:
```diff
@@ -0,0 +1,60 @@
+
+
+class LineTactic(stp.tactic.Tactic):
+    """Tactic for line up play that puts all six robots in a line on the left of the field.
```
This doc is a white lie: I never pass in the actual side of the field, even though that's a param in world_state.field. Should be a softball issue to fix whenever it becomes necessary.

kfu02 replied:
(relevant TODOs are below)

kfu02 commented on rj_gameplay/rj_gameplay/tactic/pass_tactic.py:
```diff
@@ -1,261 +1,181 @@
+
+        if self._state == "init":
+            # TODO: allow Plays to pass in this (and further below) cost fns,
+            #       otherwise behavior is not easy to manipulate
```
This is a big TODO. I'll handle this next. Having cost_fns hardcoded to the tactic was one of the reasons the old system was inflexible, so I should not have written it like this. However, to wrap my head around the FSM that handles passing, I gave simple costFns for testing: the Passer is always the robot closest to the ball, ahd the Receiver is always the robot closest to the opponent goal. In reality the play should have some leeway in deciding which robots pass between each other.

kfu02 commented on rj_gameplay/rj_gameplay/tactic/pass_tactic.py:
```diff
@@ -1,261 +1,181 @@
+            self._state = "passer_capture"
+
+        elif self._state == "passer_capture":
+            # TODO: these lines are a little ugly, any fix?
```
One idea is to have a self.passer_role and self.receiver_role, and update both after every role assignment round. But I didn't love that either. It could be much better, though, as I use this `assert isinstance` far too much in this class.

kfu02 commented on rj_gameplay/rj_gameplay/tactic/pass_tactic.py:
```diff
@@ -1,261 +1,181 @@
+            assert isinstance(receiver_role, receiver.ReceiverRole)
+
+            # TODO: create func to find good target point
+            # TODO: should update receiver_role robot every tick in the role (see skill/capture.py)
```
handled in #1824

kfu02 commented on rj_gameplay/rj_gameplay/tactic/pass_tactic.py:
```diff
@@ -1,261 +1,181 @@
+            receiver_role = self.assigned_roles[1]
+            assert isinstance(receiver_role, receiver.ReceiverRole)
+
+            # TODO: create func to find good target point
```
this is an exciting TODO for the future

kfu02 commented on rj_gameplay/rj_gameplay/tactic/wall_tactic.py:
```diff
@@ -19,120 +13,51 @@
 
+from rj_msgs.msg import RobotIntent
+
+# TODO: move this out of calculations file and into this tactic
```
handled in #1821

kfu02 commented on rj_gameplay/rj_gameplay/tactic/wall_tactic.py:
```diff
@@ -19,120 +13,51 @@
+
+        self.wall_pts = wall_calculations.find_wall_pts(self.num_wallers, world_state)
+
+        # request closest robot every pt
```
Like in PassTactic, I think the Play should have some control over what costs are passed in here, though for Wall it matters far less since "pick the closest robots" is probably desired behavior anyhow.

kfu02 commented on rj_gameplay/stp/play/__init__.py:
```diff
@@ -14,220 +15,109 @@
-
-    # mypy fails to infer that there won't be any Nones in the list.
-    return nested_results  # type: ignore
+        # TODO: use hashable Robots directly once PR #1815 merged
```
handled in #1820

kfu02 commented on rj_gameplay/stp/play/__init__.py:
```diff
@@ -14,220 +15,109 @@
+        for tactic in self.prioritized_tactics:
+            # TODO: handle if tactic requests more roles than exist
+            #       by not assigning tactic at all!
+            #       and give some default behavior (later)
```
This is an important TODO. Read design doc for more detail.

kfu02 commented on rj_gameplay/stp/role/__init__.py:
```diff
@@ -1,38 +1,48 @@
+        return self._robot
 
 
+# TODO: delete this when role assignment switched over completely
```
There are many classes below with some version of this TODO. Basically, if I delete all the old interfaces and constructs, mypy yells at me because plenty of now-obsolete classes use these constructs. We can just delete everything unused once the plays/tactics/skills have been rewritten.

kfu02 commented on rj_gameplay/stp/role/assignment/__init__.py:
```diff
@@ -1,13 +1,14 @@
 
-RoleId = Tuple[Type[tactic.ITactic], tactic.SkillEntry, int]
+# RoleId = Tuple[Type[tactic.ITactic], tactic.SkillEntry, int]
+RoleId = Any
```
RoleId is no longer used; I did this because SkillEntry no longer exists.
```suggestion
# TODO: delete once role assignment switched over entirely
RoleId = Any
```

kfu02 commented on rj_gameplay/stp/skill/__init__.py:
```diff
@@ -5,74 +5,29 @@
-        return self._registry[skill]
+    @abstractmethod
+    def is_done(self, world_state: stp.rc.WorldState) -> bool:
+        # TODO: docstring
```
```suggestion
        """True when skill is done; False otherwise.
        """
```

kfu02 reviewed (commented):
Left comments describing all of the TODOs I felt required more context. (I also wrote them assuming they are read in the context of the Files Changed tab.) Did not exhaustively explain every change: read the design doc for the general overview and read the code to see if I've done what I promised.

kfu02 committed:
update goalie role doc

kfu02 committed:
Update rj_gameplay/stp/role/assignment/__init__.py

kfu02 committed:
Update rj_gameplay/stp/skill/__init__.py

kfu02 committed:
Update rj_gameplay/rj_gameplay/gameplay_node.py

kfu02 committed:
Update rj_gameplay/rj_gameplay/role/goalie_role.py

kfu02 committed:
delete debug print

kfu02 committed:
delete unused debug print

kfu02 committed:
make pretty-lines

kfu02 committed:
fix settle skill def, add missing type imports to goalie

kfu02 committed:
fix basic style errors

kfu02 committed:
rm unused move tactic

kfu02 committed:
rm unused typing

kfu02 commented on rj_gameplay/stp/coordinator.py:
```diff
@@ -27,16 +27,14 @@ class Coordinator:
 
```
This class is now unused, can delete later

HussainGynai replied:
Why not delete now?

kfu02 committed:
final touches

kfu02 committed:
many style errors fixed

kfu02 committed:
fix final style errors (and add some ignore flags for flake8)

kfu02 committed:
actually ignore flake errors

kfu02 committed:
final final style fixes

N8BWert committed:
Tested functionality using penalty offense play -> all good

N8BWert committed:
I didn't actually call super, but it's done now

N8BWert committed:
applied superclass stuff to other skills

N8BWert committed:
Merge branch 'role-assignment-update-2022' into TODO_skill_tick_robot

kfu02 committed:
Merge pull request #1828 from RoboJackets/role-assignment-update-2022-trajectory-glitch-magic-patch

Magic fix to trajectory splicing issue

N8BWert committed:
fixed style guidelines

N8BWert committed:
Appeasing the style gods

N8BWert committed:
Style gods appeased

N8BWert reviewed (approved):
I ran all of the working plays and they all worked well.  I also discussed all of my concerns with Kevin and I support this new direction.  From a new member perspective, this direction also seems easier to understand and begin working on as it defines a more logical structure for role assignment and play selection.

p-nayak11 committed:
dribbler is default even after strp refactor

p-nayak11 committed:
Update rj_gameplay/rj_gameplay/skill/capture.py

Co-authored-by: Kevin Fu <30275369+kfu02@users.noreply.github.com>

kfu02 committed:
Merge pull request #1833 from p-nayak11/dribbler_update_strp

Dribbler update strp

Yudai-8-om committed:
Merge pull request #1820 from RoboJackets/apply_hash_robot

Apply hashable robot

gordonm867 committed:
Delete trajectory test that no longer works

Because of recent changes to prevent a crash from trajectory splicing in the settle planner, there is a test here that is no longer applicable to our current code.  For now, we have commented it out.

oddlyspookycherry committed:
merge fix

oddlyspookycherry committed:
repr for skills

kfu02 committed:
Merge branch 'ros2' into role-assignment-update-2022

kfu02 committed:
fix style errors from merges

kfu02 committed:
Merge branch 'ros2' into role-assignment-update-2022

kfu02 committed:
fix nmark mypy merge issues

kfu02 committed:
make pretty

gordonm867 committed:
"Ran make pretty-lines"

gordonm867 committed:
Merge pull request #1836 from RoboJackets/if-the-test-does-not-work-then-just-delete-it

Delete trajectory test that no longer works

N8BWert committed:
move checks for None robot. But encountered error in role assignment.

N8BWert committed:
I thought this file was added but it wasn't

N8BWert committed:
moved everything back to superclass

N8BWert committed:
Updated comments

N8BWert committed:
Merge pull request #1824 from RoboJackets/TODO_skill_tick_robot

self.robot is updated in Skill superclass tick method

HussainGynai commented on rj_gameplay/rj_gameplay/gameplay_node.py:
```diff
@@ -30,27 +28,14 @@
-from rj_gameplay.play import penalty_defense
+
+# ignore "unused import" error
+from rj_gameplay.play import line_up, basic_defense, keepaway  # noqa: F401
```
What's the comment here mean?

kfu02 replied:
pylint complains that we have unused imports on this line without the `# noqa` flag

kfu02 replied:
but imo it's fine to have unused imports on this specific line because test plays change so often, and that's what this is for

kfu02 replied:
OH also I wanted people to be able to see which plays have been rewritten thus far, so I imported all even if they aren't all in use

HussainGynai commented on rj_gameplay/stp/play/__init__.py:
```diff
@@ -14,215 +15,109 @@
+                    if robot in used_robots:
+                        continue
+                    if not robot.visible:
+                        continue
```
Make these 2 one if statement

HussainGynai replied:
Smh didn't even change this

HussainGynai commented on rj_gameplay/stp/role/cost.py:
```diff
@@ -1,83 +1,60 @@
+        return 1e9
+
+    def __repr__(self):
+        return f"PickClosestRobot(target_point={self._target_point})"
```
Not really that important, but maybe one for closest to ball or something could be useful too

kfu02 replied:
you can set target to world_state.ball.pos for that. I'll change the costFn name to PickClosestToPoint to make that more obvious

HussainGynai replied:
Yeah but the ball can move, and cost functions already take in world state.

HussainGynai commented on rj_gameplay/stp/tactic/__init__.py:
```diff
@@ -1,239 +1,68 @@
-    """A dictionary mapping typed keys to a list of skill.ISkill."""
+RoleRequests = Any
+RoleResults = Any
+SkillEntry = Any
```
Are these three types used?

kfu02 replied:
No, left in for legacy gameplay classes to not trigger mypy errors. Will make a cleanup PR that deletes unused/legacy code

HussainGynai commented on rj_gameplay/stp/tactic/__init__.py:
```diff
@@ -1,239 +1,68 @@
+    def __init__(self, world_state: stp.rc.WorldState) -> None:
+        """Save the world_state, and create empty lists for handling role requests."""
 
+        self.world_state = world_state
```
Do we need to save the world_state? It should already be given on every tick.

kfu02 replied:
That is very true. Will delete

HussainGynai commented on rj_gameplay/rj_gameplay/role/passer.py:
```diff
@@ -0,0 +1,69 @@
+        if self._state == "init":
+            self.receive_skill = receive.Receive(robot=self.robot)
+            intent = self.receive_skill.tick(world_state)
+            self._state = "receiving"
```
Why does the pass start by receiving? Is this supposed to be capture, might want to name it better for less confusion.

kfu02 replied:
Bad name; pass starts by capturing the ball, I named it "receiving" b/c the skill is named Receive. Will change

HussainGynai commented on rj_gameplay/rj_gameplay/play/defensive_clear.py:
```diff
@@ -11,7 +11,6 @@
 from typing import Dict, List, Tuple
```
Is this file still gonna be a thing? If so add a TODO or issue to make it a play in the new system

kfu02 replied:
It will be. All plays besides BasicDefense and LineUp are old, so I didn't bother commenting above every old STP "TODO: rewrite me".

HussainGynai commented on rj_gameplay/rj_gameplay/play/line_up.py:
```diff
@@ -1,75 +1,47 @@
+                self.state = "done"
+            return self.get_robot_intents(world_state)
+        elif self.state == "done":
+            # TODO: does this state need to exist?
```
All plays should probably have a shared "done" flag or something so that play selection knows it needs to pick a new play.

kfu02 replied:
Ongoing debate whether one situation = one play or if situations should have the ability to toggle plays. I'm of the 1:1 camp which is why I left this comment (since if 1:1 the situation will end before the play does, thereby switching the play)

HussainGynai commented on rj_gameplay/rj_gameplay/role/passer.py:
```diff
@@ -0,0 +1,69 @@
+        # this state transition is done by the PassTactic, which is not canonical FSM
+        elif self._state == "init_execute_pass":
+            # TODO: make these params configurable
+            self.pivot_kick_skill = pivot_kick.PivotKick(
```
This needs to be able to take in a new target point every tick, or we need someway to change the target point during pivot kick.

kfu02 replied:
Fair, need to split PivotKick into Pivot and Kick for this. Will change

HussainGynai commented on rj_gameplay/rj_gameplay/skill/pivot_kick.py:
```diff
@@ -54,16 +55,30 @@ def __init__(
-            return self.pivot.tick(robot, world_state, intent)
+        self._state = "capture"
+
+    def tick(self, world_state: rc.WorldState) -> RobotIntent:
```
Needs a way to change the target point on every tick

kfu02 replied:
see above reply

HussainGynai commented on rj_gameplay/rj_gameplay/tactic/pass_tactic.py:
```diff
@@ -1,261 +1,183 @@
+        elif self._state == "passer_capture":
+            # TODO: these lines are a little ugly, any fix?
+            passer_role = self.assigned_roles[0]
+            assert isinstance(passer_role, passer.PasserRole)
```
Is this for debug?

HussainGynai commented on rj_gameplay/rj_gameplay/tactic/pass_tactic.py:
```diff
@@ -1,261 +1,183 @@
+            if passer_role.pass_ready:
+                self._state = "get_receiver"
+
+        elif self._state == "get_receiver":
```
Is the passer pivoting in this step? I feel like the getting receiver step should be something that finishes right before the pass is executed.

HussainGynai commented on rj_gameplay/stp/skill/__init__.py:
```diff
@@ -5,74 +5,32 @@
+from rj_msgs.msg import RobotIntent
 
 
 class ISkill(ABC):
```
Is there any need for theses interfaces, if not we can delete all of them.

kfu02 replied:
they exist so legacy code doesn't brick the compiler, same as RoleRequest et al

HussainGynai reviewed (changes requested):
Some minor changes, except for passing stuff. I think stuff like getting the receiver and passing should happen concurrently pretty much until the passer physically kicks. Part of this the passer should also be able to change its passing location each tick. Also change the strings to enums for states.

kfu02 committed:
rm coordinator class, clarify closest-to-pt costfn by rename, rm superfluous line in tactic init, change confusing passer fsm state name

kfu02 committed:
merge ros2, reformat

kfu02 committed:
explain blank move tactic file, make state private with leading _

kfu02 committed:
explain blank move tactic file, make state private with leading _

kfu02 committed:
enum passer, receiver, piv kick

kfu02 committed:
enum basic def, pass tac, rm debug piv kick print

kfu02 committed:
line up and keepaway enumd

kfu02 committed:
Merge branch 'role-assignment-update-2022' of https://github.com/RoboJackets/robocup-software into role-assignment-update-2022

kfu02 committed:
rm asserts

kfu02 commented:
@HussainGynai I addressed most of your concerns in commits, except I put state enums here: #1843 

For passing, I think it's very likely we need more flexibility than what we currently have, but that we need to test more to be certain exactly what changes are needed. For instance, after testing, we could conclude that pivot kick is a fast enough action to where the target point doesn't need to change every tick. (Again, unlikely, but you never know). Regardless, I believe that perfecting passing is out of the scope of this PR, so I made a [new issue for it](https://app.clickup.com/t/2a0zk1p). If we wait to merge this PR until passing works 100%, this branch will become the new `ros2`, in that everyone treats it as the defacto `main`.

For unit testing (which you mentioned to me IRL but not here), I'm not sure what to unit test? There's not much useful output that can be seen without the full simulated environment because it's infrastructure dependent on world_state and external path planners. The best I can think is to test that the superclasses work, which is what the few concrete classes I chose to add already do. Maybe the role assignment method in `stp/play/__init__.py` is a good candidate, but seeing as that's also unfinished I would prefer to do it after a default unassigned Role is added.

kfu02 committed:
Merge branch 'role-assignment-update-2022' into state-enums

oddlyspookycherry committed:
final changes

oddlyspookycherry committed:
final changes

oddlyspookycherry committed:
style fix

oddlyspookycherry committed:
style fix

oddlyspookycherry committed:
style fix

oddlyspookycherry committed:
final checkout

oddlyspookycherry committed:
Merge pull request #1837 from RoboJackets/repr-skills

Repr for skills

kfu02 committed:
rm done todos

HussainGynai reviewed (approved):
LGTM, just change the if statement thing

kfu02 committed:
Merge pull request #1843 from RoboJackets/state-enums

Make FSMs in gameplay Enum-based
